common/lanelet2_extension/lib/SignalizedIntersection.cpp

```diff
@@ -0,0 +1,139 @@
+                                                        entry.end());
+        rules[lanelet::CarmaRoleNameString::IntersectionExit].insert(rules[lanelet::CarmaRoleNameString::IntersectionExit].end(), exit.begin(),
+                                                        exit.end());
+        rules[lanelet::CarmaRoleNameString::IntersectionInterior].insert(rules[lanelet::CarmaRoleNameString::IntersectionInterior].end(), interior.begin(),
+                                                        interior.end());
+
+        // Add attributes
+        AttributeMap attribute_map({ { AttributeNamesString::Type, AttributeValueString::RegulatoryElement },
+                                    { AttributeNamesString::Subtype, RuleName }});
+
+
+        return std::make_unique<RegulatoryElementData>(id, rules, attribute_map);
+    }
+
+    std::vector<CarmaTrafficLightConstPtr> SignalizedIntersection::getTrafficSignals(const ConstLanelet& llt) const
+    {
+        return llt.regulatoryElementsAs<CarmaTrafficLight>();
+    }
+
+    void SignalizedIntersection::addLanelet(const Lanelet& lanelet, IntersectionSection section)
+    {
+        switch (section)
+        {
+            case IntersectionSection::ENTRY:
+                parameters()[RoleNameString::Refers].emplace_back(RuleParameter(lanelet));
+                break;
+            case IntersectionSection::EXIT:
+                parameters()[CarmaRoleNameString::IntersectionExit].emplace_back(RuleParameter(lanelet));
+                break;
+            case IntersectionSection::INTERIOR:
+                parameters()[CarmaRoleNameString::IntersectionInterior].emplace_back(RuleParameter(lanelet));
+                break;
+            default:
+                throw std::invalid_argument("Invalid section is passed as IntersectionSection");
+        }
+    }
+    
+    namespace {
+
+        template <typename T>
+        bool findAndErase(const T& primitive, RuleParameters* member)
+        {
+        if (member == nullptr)
+        {
+            std::cerr << __FUNCTION__ << ": member is null pointer";
+            return false;
+        }
+        auto it = std::find(member->begin(), member->end(), RuleParameter(primitive));
+        if (it == member->end())
+        {
+            return false;
+        }
+        member->erase(it);
+        return true;
+        }
+    }
+
+    bool SignalizedIntersection::removeLanelet(const Lanelet& llt)
+    {
+        // if successfully found and erased, then return true, else false
+        if (findAndErase(RuleParameter(llt), &parameters().find(RoleNameString::Refers)->second))   
+            return true;
+        if (findAndErase(RuleParameter(llt), &parameters().find(CarmaRoleNameString::IntersectionExit)->second))
+            return true;
+        if (findAndErase(RuleParameter(llt), &parameters().find(CarmaRoleNameString::IntersectionInterior)->second))
+            return true;
```

Comment:
This logic looks like it checks this lanelet against being entry, exist, and interior. Is there no way to check the role first then remove it instead of calling find 3 times?

Reply:
I have included a map<lanelet::Id, section> so now it only does once instead of three. I think this will also prove useful when dealing with IntersectionManager in the near future.